File: frontend/src/types/questionnaire.ts
/**
 * Questionnaire agent types matching the backend models in
 * backend/app/models/questionnaire.py
 */

/** Returned when the agent wants to ask the user something. */
export interface AgentQuestion {
  session_id: string;
  type: "question";
  question: string;
  context?: string | null;
  options?: string[] | null;
}

/** A single generated compliance question. */
export interface GeneratedQuestion {
  id: string;
  question: string;
  category: string;
  priority: string;
  expected_evidence?: string | null;
  guidance_notes?: string | null;
}

/** Questions generated for one framework control. */
export interface ControlQuestions {
  control_id: string;
  control_title: string;
  framework: string;
  questions: GeneratedQuestion[];
}

/** Returned when the agent finishes generating all questions. */
export interface QuestionnaireComplete {
  session_id: string;
  type: "complete";
  controls: ControlQuestions[];
  total_controls: number;
  total_questions: number;
  generation_time_ms: number;
  criteria_summary: string;
}

/** Discriminated union — narrows on `type` field. */
export type QuestionnaireResponse = AgentQuestion | QuestionnaireComplete;

/** Structured criteria for the wizard flow (batch generation). */
export interface GenerateWithCriteriaRequest {
  project_id: string;
  assessment_id?: string | null;
  maturity_level: "first_time_audit" | "recurring_assessment" | "mature_isms";
  question_depth: "high_level_overview" | "balanced" | "detailed_technical";
  priority_domains: string[];
  compliance_concerns?: string | null;
  controls_to_skip?: string | null;
}

/** Summary row from GET /questionnaire/sessions. */
export interface QuestionnaireSessionSummary {
  id: string;
  status: string;
  total_questions: number;
  total_controls: number;
  created_at: string;
  assessment_id: string | null;
}

/** Full session row from GET /questionnaire/sessions/:id. */
export interface QuestionnaireSessionDetail {
  id: string;
  project_id: string;
  client_id: string;
  status: string;
  assessment_id: string | null;
  generated_questions: ControlQuestions[];
  total_controls: number;
  total_questions: number;
  generation_time_ms: number;
  agent_criteria: { summary: string };
  created_at: string;
}

File: backend/app/models/questionnaire.py
"""Questionnaire agent request and response models."""

from typing import Literal, Optional

from pydantic import BaseModel, Field


class GenerateWithCriteriaRequest(BaseModel):
    """Structured criteria for batch question generation (wizard flow)."""

    project_id: str = Field(..., description="Project UUID")
    assessment_id: Optional[str] = Field(
        None, description="Assessment UUID to link generated questions to"
    )

    # Step 1: Assessment Profile
    maturity_level: str = Field(
        ..., description="Organization's assessment maturity"
    )
    question_depth: str = Field(
        ..., description="Desired question depth/count per control"
    )

    # Step 2: Focus & Priorities
    priority_domains: list[str] = Field(
        default_factory=list,
        description="Control domains to emphasize (empty = all equal)",
    )
    compliance_concerns: Optional[str] = Field(
        None, description="Known gaps or specific concerns"
    )
    controls_to_skip: Optional[str] = Field(
        None, description="Controls to de-emphasize or skip"
    )
    questions_per_control: Optional[int] = Field(
        None, description="Override questions per control (default 3)"
    )


class GenerateQuestionRequest(BaseModel):
    """Request to start a new questionnaire generation session."""

    project_id: str = Field(..., description="Project UUID")
    assessment_id: Optional[str] = Field(
        None, description="Assessment UUID to link generated questions to"
    )


class RespondRequest(BaseModel):
    """Request to continue a session by answering the agent's question."""

    session_id: str = Field(
        ..., description="Session UUID from generate-question"
    )
    answer: str = Field(
        ..., min_length=1, description="User's answer to the agent's question"
    )


class AgentQuestion(BaseModel):
    """Returned when the agent wants to ask the user something."""

    session_id: str
    type: Literal["question"] = "question"
    question: str
    context: Optional[str] = None
    options: Optional[list[str]] = None


class GeneratedQuestion(BaseModel):
    """A single generated compliance question."""

    id: str
    question: str
    category: str
    priority: str
    expected_evidence: Optional[str] = None
    guidance_notes: Optional[str] = None


class ControlQuestions(BaseModel):
    """Questions generated for one framework control."""

    control_id: str
    control_title: str
    framework: str
    questions: list[GeneratedQuestion]


class QuestionnaireComplete(BaseModel):
    """Returned when the agent finishes generating all questions."""

    session_id: str
    type: Literal["complete"] = "complete"
    controls: list[ControlQuestions]
    total_controls: int
    total_questions: int
    generation_time_ms: int
    criteria_summary: str


class GenerationRedirect(BaseModel):
    """Returned when the conversational agent triggers batch generation."""

    session_id: str
    type: Literal["generation_redirect"] = "generation_redirect"
    criteria: dict


# Union response type
QuestionnaireResponse = AgentQuestion | QuestionnaireComplete | GenerationRedirect
